### ISSUE TITLE
Generate both a LIVE and EVENT HLS .m3u8 for a live stream

### ISSUE THREAD

**Issue by gregmsanderson** [google/shaka-packager#849]
I see in the docs that you can use the flag to make either a LIVE or EVENT m3u8 for a live stream:
`--hls_playlist_type <type>`

However I'd like both.

Why? Well during the live stream, the 'LIVE' type is better. As that acts like a sliding window, only listing the last X segments, meaning the .m3u8 is super small.

But when the live stream _ends_, I'd like to make a VOD of it. Armed with an existing EVENT .m3u8 (which would be there behind the scenes) that is super-easy to do. As it lists all the segments.

So ... I'd like Shaka packager to output _both_ a live.m3u8 and event.m3u8. Is that possible? Or do I have to run two instances of it?

Thanks.

**Comment by gregmsanderson** [comment]
Hmm it _seems_ it's not possible given the code, understandably, expects a single sub-m3u8 type. And based on that, sets the header and decides whether to remove segments from it. e.g

https://github.com/google/shaka-packager/blob/master/packager/hls/base/media_playlist.cc#L127
https://github.com/google/shaka-packager/blob/master/packager/hls/base/media_playlist.cc#L663

So if you did want to add this feature, I guess a new value would be needed. Like. `'live,event'`
`--hls_playlist_type live,event`
... to mean output both LIVE and EVENT, and adjusting the switch () etc accordingly to output two sub m3u8 files. Each working as now. Just, two of them.

I can't see a simple way to make a VOD from a live stream without this ability unless you output EVENT to begin with. Which would also solve the problem. But for a long stream, that means a huge .m3u8 being requested over and over again. So I do think this ability would be useful.

**Comment by kqyang** [comment]
> Hmm it seems it's not possible given the code, understandably, expects a single sub-m3u8 type.

Correct.

> I can't see a simple way to make a VOD from a live stream without this ability unless you output EVENT to begin with. 

I think it is possible to create a VOD from a live stream with LIVE playlist type. 

What you can do with the LIVE playlist is: (1) Make sure the buffer depth of the live stream (`--time_shift_buffer_depth`) covers the duration of the stream that you want to turn to VOD; (2) Prepare a script that copies all the media segments and playlists; (3) When it is time to create the VOD, execute the script, then modifies the playlists for VOD.

It does require users to run the scripts multiple times manually to create multiple VODs.

> So if you did want to add this feature, I guess a new value would be needed. Like. 'live,event'

Yes, that will be one way to do it. A simple way to implement it is to create two separate HLS Notifiers, one for live and one for event (though that is not the most efficient way to do it). 

That being said, I am not sure if it is a good idea as the event playlist will grow along with the increase in the live stream duration.

It is probably better to build what is described above on how to turn live to VOD into packager or a separate automated script. One challenge is when to create the VODs - maybe periodically? That certainly requires some more work.

**Comment by gregmsanderson** [comment]
Thanks. That idea, using the buffer depth, would indeed work _if_ you always know the length before the event starts. _But_ for user-generated content, that won't be the case at the server end. And even the user may not know how long their own event may end up lasting. Event overruns and such. Rather than enforce a cut-off time, set at the start, I'd like a simple 'End' whenever it ends.

Hence needing an EVENT m3u8 as a sibling of the LIVE m3u8 which has all the segments in, however long the stream lasts.

(Yes, that EVENT .m3u8 could be built dynamically just from a folder of .ts segment files. But that is awkward, as the last segment's duration would not be known. And so have to delve into ffprobe etc.)

Looking at Twitch, it seems this is what they do, as a VOD .m3u8 is instantly available when a stream ends. And it's different to the sliding-window LIVE .m3u8 during the stream. So I assume they are generating _both_ .m3u8 at the same time.

> That being said, I am not sure if it is a good idea as the event playlist will grow along with the increase in the live stream duration.

But that's fine, because that .m3u8 won't be served until the event ends. And a VOD .m3u8 is always going to be huge for a long video when listing separate segments. It's ok for a VOD though as it's only requested once. Unlike a live stream where it's requested every X seconds.

> One challenge is when to create the VODs - maybe periodically?

Ah, well that sounds like you are thinking about creating VODs from a single stream, like once an hour etc. That indeed would make things complicated. All I'm talking about is making one VOD. At the end. Basically by adding an ENDLIST. And changing EVENT to VOD. As that is pretty much all there is too it. Assuming the EVENT .m3u8 is there to begin with.

When you talk about HLS notifiers, is that something easy to change? The question for me is whether that is more efficient than running two packagers, one making EVENT, and one making LIVE. And using the LIVE during the event. And using the EVENT after the event (via the changes).

**Comment by kqyang** [comment]
> When you talk about HLS notifiers, is that something easy to change?

Yes, it will be a relatively quick change. You are welcomed to work on it. You can modify [MuxerListenerFactory](https://github.com/google/shaka-packager/blob/a1dd82d478c4622dff1a0d0489f170ed6e1814d5/packager/media/event/muxer_listener_factory.cc) to accept an additional `HlsNotifier`, then modify the callers.

> Hence needing an EVENT m3u8 as a sibling of the LIVE m3u8 which has all the segments in, however long the stream lasts.

It is ok as long as it is not too "long", otherwise you'll see performance problems and storage problems.

> The question for me is whether that is more efficient than running two packagers, one making EVENT, and one making LIVE. 

It is definitely more efficient then running two packagers as the media streams (which is the time consuming part) are only created once.

**Comment by gregmsanderson** [comment]
Interesting ...

Ok, thanks.
